sandbox/libs/authn/src/main/java/org/opensearch/authn/Permissions.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.authn;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * An extension of the abstract Permission class which uses String-object Permissions.
+ *
+ * Example "opensearch.indexing.index.create"
+ *
+ * @opensearch.experimental
+ */
+public class Permissions extends Permission {
+
+    private final static String PERMISSION_DELIMITER = "\\.";
+
+    public final static String[] INVALID_CHARACTERS = new String[] { ":", "" }; // This is a placeholder for what may want to be banned
+
+    // This is a placeholder for the different actions which a permission may grant a Subject to perform
+    public final static String[] QUALIFIED_ACTIONS = new String[] { "CREATE", "READ", "WRITE", "DELETE", "UPDATE" };
+
+    // A placeholder for the different resources which a permission may grant a permission based on
+    public final static String[] QUALIFIED_RESOURCES = new String[] { "index", "indices", "cluster", "all" };
+
+    public String permissionString;
+
+    public void Permission(String permission) {
+        this.permissionString = permission;
+    }
+
+    /**
+     * Check that the permission does not contain any forbidden strings.
+     * This set implementation does so in O(n).
+     * Assumes that the permission is formatted as <principal>.<resource>.<action>
+     * The principal should already be verified before the permission is created.
+     */
+    public boolean permissionIsValidFormat() {
```

Comment:
computationally this string manipulation is kind of expensive, what do you think about changing the storage of the permissionString into its composite parts?

Reply:
That sounds good to me and would let you do just 1:1 comparison with the same time complexity as the set logic. Great call!